### ISSUE TITLE
batch: Push limit to table scan.

### ISSUE THREAD

**Issue by liurenjie1024** [singularity-data/risingwave#3481]
For following sql:
```
select * from t order by a limit 10000, 10
```

And we assume that t is ordered by a in each patition, we should generate following plan:
```
   TopN(offset = 10000, limit = 10)
          |
    MergeSortExchange
            |
    TableScan(limit = 10010)
```

This way we don't need to pull the first 10000 rows into executor, which can avoid encoding/decoding of those rows. But the storage layer need to provide api for those skipping.

**Comment by fuyufjh** [comment]
It sounds like early-stop of `TableScan` rather than "Push limit to table scan".

**Comment by fuyufjh** [comment]
Well... IIUC, in the iterator API, you may just call `next()` without reading (deserializing) the data.

**Comment by liurenjie1024** [comment]
> Well... IIUC, in the iterator API, you may just call `next()` without reading (deserializing) the data. An explicit `skip` is not necessray.

Do you mean the `TableIter` api?  It seems that its provides iterator on rows and data chunk, so I'm afraid we can't avoid decoding data.

**Comment by xxchan** [comment]
How to do this when there are parallel TableScan? What's the `offset` for each executor?  👀

**Comment by liurenjie1024** [comment]
> How to do this when there are parallel TableScan? What's the `offset` for each executor? 👀

The `offset` is valid for each partition.

**Comment by xxchan** [comment]
> The `offset` is valid for each partition.

Still didn't get it... 

e.g., `t` has data 1~20000, one executor scans 1~10000, and another scans 10000~20000. The results should be the first 10 rows of the second scan?

**Comment by liurenjie1024** [comment]
> > The `offset` is valid for each partition.
> 
> Still didn't get it...
> 
> e.g., `t` has data 1 ~ 20000, one executor scans 1 ~ 10000, and another scans 10000 ~ 20000. The results should be the first 10 rows of the second one? So the second one should not limit 10000?

I see, there are some problem with problem statement. We should not skip in table scan, for the sql in issue, we should scan 10010 for each partition.

Your example is a corner case, and we should scan all data for each partition.

Maybe we just need to put a limit operator above table scan or add a limit option in table scan.

**Comment by xxchan** [comment]
> Even if the limit/offset were pushed down by optimizer, there are still many rows that can be avoid from being scanned.

@fuyufjh How to scan fewer rows? I think each partition have to scan limit+offset rows before topn can produce results?

**Comment by fuyufjh** [comment]
> How to scan fewer rows? I think each partition have to scan limit+offset rows before topn can produce results?

Yes. Ignore me.

**Comment by fuyufjh** [comment]
So how about this idea:

> It sounds like early-stop of TableScan rather than "Push limit to table scan".

I don't think it's necessary to put a `Limit` above `TableScan`

**Comment by hzxa21** [comment]
> > It sounds like early-stop of `TableScan` rather than "Push limit to table scan".
> > Even if the limit/offset were pushed down by optimizer, there are still many rows that can be avoid from being scanned. To accomplish this, the TableScan executor should be stopped and dropped as soon as the upper `TopN` has gathered enough rows.
> 
> In fact, I'm expecting the storage team to provide something to skip `offset` rows, which avoids fetching and decoding of rows. cc @hzxa21

The KV storage doesn't understand rows so it cannot simply skip on a row basis. It is possible to let the relation table layer to do the skipping but under the hoop it just calls next for `offset` times and ignore the result so it won't save anything.

**Comment by xxchan** [comment]
Then we should add limit to MergeSortExchange?

**Comment by liurenjie1024** [comment]
> Then we should add limit to MergeSortExchange?

Maybe not, as @fuyufjh said, top Limit operator can stop pulling rows. 

Seems that we have discussed this before, and we can close this for now.

**Comment by xxchan** [comment]
Limit can stop early, but TopN cannot?  🤔

**Comment by liurenjie1024** [comment]
> Limit can stop early, but TopN cannot yet? 🤔

I think TopN can do same thing.

**Comment by xxchan** [comment]
> I think TopN can do same thing.

TopN needs to sort first before it can emit results (and maybe exit early by `limit`). It consumes all data from the child, and cannot early exit.


There's no early exit in the code:

```rust
impl TopNExecutor {
    #[try_stream(boxed, ok = DataChunk, error = RwError)]
    async fn do_execute(mut self: Box<Self>) {
        #[for_await]
        for data_chunk in self.child.execute() {
            let data_chunk = data_chunk?;
            self.top_n_heap.fit(data_chunk);
        }

        if let Some(data_chunk) = self.top_n_heap.dump(self.offset) {
            let batch_chunks = DataChunk::rechunk(&[data_chunk], DEFAULT_CHUNK_BUFFER_SIZE)?;
            for ret_chunk in batch_chunks {
                yield ret_chunk
            }
        }
    }
}
```

**Comment by liurenjie1024** [comment]
> > I think TopN can do same thing.
> 
> TopN needs to sort first before it can emit results (and maybe exit early by `limit`). It consumes all data from the child, and cannot early exit.
> 
> There's no early exit in the code:
> 
> ```rust
> impl TopNExecutor {
>     #[try_stream(boxed, ok = DataChunk, error = RwError)]
>     async fn do_execute(mut self: Box<Self>) {
>         #[for_await]
>         for data_chunk in self.child.execute() {
>             let data_chunk = data_chunk?;
>             self.top_n_heap.fit(data_chunk);
>         }
> 
>         if let Some(data_chunk) = self.top_n_heap.dump(self.offset) {
>             let batch_chunks = DataChunk::rechunk(&[data_chunk], DEFAULT_CHUNK_BUFFER_SIZE)?;
>             for ret_chunk in batch_chunks {
>                 yield ret_chunk
>             }
>         }
>     }
> }
> ```

Yes, you are right. TopN doesn't assume its inputs are sorted. This case I think it would be better to push limit as low as possible to further reduce scan.

**Comment by xxchan** [comment]
TLDR: 
- For Limit
  - ❌pushing `offset` doesn't have performance gain according to storage. 
  - ❌pushing `limit` is unnecessary since Limit can early exits.
- For TopN
  - ❌pushing `offset` is incorrect. 
  - pushing `limit+offset` as a limit (to each source of MergeSortExchange, not necessarily TableScan?) is an optimization. Btw I think this limit can early exit, so we do not need to push deeper?

**Comment by liurenjie1024** [comment]
Another rule for simplifying the plan is that if the input of TopN is already sorted on the key, we can convert TopN to Limit, e.g.:

```
TopN(limit = 10000, sort = a)                                                  Limit(limit = 10000)
    |                                                               =>                           |
 TableScan(sort = a)                                                                 TableScan(sort = a) 
```

This also applieds to `LocalTopN`

**Comment by st1page** [comment]
> Another rule for simplifying the plan is that if the input of TopN is already sorted on the key, we can convert TopN to Limit, e.g.:
> 
> ```
> TopN(limit = 10000, sort = a)                                                  Limit(limit = 10000)
>     |                                                               =>                           |
>  TableScan(sort = a)                                                                 TableScan(sort = a) 
> ```
> 
> This also applieds to `LocalTopN`

and we can eliminate the max and min, rewrite the plan to limit too. https://docs.pingcap.com/tidb/stable/max-min-eliminate

**Comment by lmatz** [comment]
`Limit` doing early exist still suffers a little 
as `limit` may be larger than the chunk size `Scan` intended to get from storage?
